packages/core/src/telemetry/config.ts

```diff
@@ -34,11 +34,237 @@ export function parseTelemetryTargetValue(
+  if (
+    (value.startsWith('"') && value.endsWith('"')) ||
+    (value.startsWith("'") && value.endsWith("'"))
+  ) {
+    return value.slice(1, -1).replace(/\\(.)/g, '$1');
+  }
+  return value;
+};
+
+function splitByDelimitersRespectingQuotes(input: string): string[] {
+  const parts: string[] = [];
+  let currentPart = '';
+  let state = createInitialState();
+
+  for (const char of input) {
+    state = updateParserState(state, char);
+
+    if (isUnquotedDelimiter(char, state)) {
+      const trimmed = currentPart.trim();
+      if (trimmed) {
+        parts.push(trimmed);
+      }
+      currentPart = '';
+    } else {
+      currentPart += char;
+    }
+  }
+
+  if (state.inQuotes) {
+    throw new Error(
+      `Unclosed quote in header value. Expected closing ${state.quoteChar}`,
+    );
+  }
+
+  const trimmed = currentPart.trim();
+  if (trimmed) {
+    parts.push(trimmed);
+  }
+
+  return parts;
+}
+
+function parseKeyValuePair(
+  pair: string,
+): { key: string; value: string } | null {
+  const equalsIndex = pair.indexOf('=');
+  if (equalsIndex === -1) {
+    return null;
+  }
+
+  const key = pair.substring(0, equalsIndex).trim();
+  const rawValue = pair.substring(equalsIndex + 1).trim();
+
+  if (!isValidHeaderName(key)) {
+    return null;
+  }
+
+  if (
+    (rawValue.startsWith('"') && !rawValue.endsWith('"')) ||
+    (rawValue.startsWith("'") && !rawValue.endsWith("'"))
+  ) {
+    throw new Error('Unclosed quote in header value');
+  }
+
+  const value = unquote(rawValue);
+
+  if (!key || !value || !isValidHeaderValue(value)) {
+    return null;
+  }
+
+  return { key, value };
+}
+
+function countUnquotedEquals(input: string): number {
+  let state = createInitialState();
+  let count = 0;
+
+  for (const char of input) {
+    state = updateParserState(state, char);
+    if (isUnquotedEquals(char, state)) {
+      count++;
+    }
+  }
+
+  return count;
+}
+
+/**
+ * Parse OTLP headers from a string value.
+ * Supports JSON format or key=value pairs separated by comma/semicolon.
+ * Validates per RFC 7230 and enforces reasonable limits to prevent
+ * accidental misuse and protect OTLP backends.
+ */
+export function parseOtlpHeaders(
+  value: string | undefined,
+): Record<string, string> | undefined {
+  if (value === undefined || value.trim() === '') {
+    return undefined;
+  }
+
+  const trimmedValue = value.trim();
+
+  if (trimmedValue.startsWith('{')) {
+    try {
+      const parsed = JSON.parse(trimmedValue);
+      if (
+        typeof parsed === 'object' &&
+        parsed !== null &&
+        !Array.isArray(parsed)
+      ) {
+        const result: Record<string, string> = {};
+        for (const [key, val] of Object.entries(parsed)) {
+          if (!key || val === undefined || val === null) {
+            continue;
+          }
+          const stringVal = String(val);
+          if (isValidHeaderName(key) && isValidHeaderValue(stringVal)) {
+            result[key] = stringVal;
```

Comment:
see the other comment about duplicated header behavior. I'm also just generally worried about parsing headers, which is surprising complicated. look at the issues on this NPM package, for example: https://github.com/kesla/parse-headers